## Data Cleaning

In this notebook, I perform a preliminary review of the dataset using pands profiling report, then clean accordingly based on results of review. 

### Contents
- [Import libraries and data](#Import-libraries-and-data) 
- [Preliminary review of data](#Detailed-review)
- [Deal with missing values](#Dealing-with-missing-values)
- [Save out data](#Save-cleaned-data)

### Import libraries and data

In [74]:
# import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pandas_profiling import ProfileReport
import seaborn as sns
plt.style.use('seaborn-white')
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',150)
%matplotlib inline

In [75]:
# read data
train = pd.read_csv("../data/train.csv")
test = pd.read_csv("../data/test.csv")
# preview data
print("Top 5 rows of train data".center(50,"="))
display(train.head())
print()
print(f"Shape of train data : {train.shape}")
print()
print("Data types in train data".center(50,"="))
print(train.dtypes)
print()
print()
print("Top 5 rows of test data".center(50,"="))
display(test.head())
print()
print(f"Shape of test data : {test.shape}")
print()
print("Data types in test data".center(50,"="))
print(test.dtypes)

=============Top 5 rows of train data=============


Id  Product_Info_1 Product_Info_2  Product_Info_3  Product_Info_4  \
0   2               1             D3              10        0.076923   
1   5               1             A1              26        0.076923   
2   6               1             E1              26        0.076923   
3   7               1             D4              10        0.487179   
4   8               1             D2              26        0.230769   

   Product_Info_5  Product_Info_6  Product_Info_7   Ins_Age        Ht  \
0               2               1               1  0.641791  0.581818   
1               2               3               1  0.059701  0.600000   
2               2               3               1  0.029851  0.745455   
3               2               3               1  0.164179  0.672727   
4               2               3               1  0.417910  0.654545   

         Wt       BMI  Employment_Info_1  Employment_Info_2  \
0  0.148536  0.323008              0.028                 12   
1  0.131799  0.272288              0.000                  1   
2  0.288703  0.428780              0.030                  9   
3  0.205021  0.352438              0.042                  9   
4  0.234310  0.424046              0.027                  9   

   Employment_Info_3  Employment_Info_4  Employment_Info_5  Employment_Info_6  \
0                  1                0.0                  3                NaN   
1                  3                0.0                  2             0.0018   
2                  1                0.0                  2             0.0300   
3                  1                0.0                  3             0.2000   
4                  1                0.0                  2             0.0500   

   InsuredInfo_1  InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0              1              2              6              3              1   
1              1              2              6              3              1   
2              1              2              8              3              1   
3              2              2              8              3              1   
4              1              2              6              3              1   

   InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
0              2              1                    1                    1   
1              2              1                    2                    1   
2              1              1                    2                    1   
3              2              1                    2                    1   
4              2              1                    2                    1   

   Insurance_History_3  Insurance_History_4  Insurance_History_5  \
0                    3                    1             0.000667   
1                    3                    1             0.000133   
2                    1                    3                  NaN   
3                    1                    3                  NaN   
4                    1                    3                  NaN   

   Insurance_History_7  Insurance_History_8  Insurance_History_9  \
0                    1                    1                    2   
1                    1                    3                    2   
2                    3                    2                    3   
3                    3                    2                    3   
4                    3                    2                    3   

   Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  Family_Hist_5  \
0              2            NaN       0.598039            NaN       0.526786   
1              2       0.188406            NaN       0.084507            NaN   
2              3       0.304348            NaN       0.225352            NaN   
3              3       0.420290            NaN       0.352113            NaN   
4              2       0.463768            NaN       0.408451            NaN   

   Medical_History_1  


Shape of train data : (59381, 128)

=============Data types in train data=============
Id                       int64
Product_Info_1           int64
Product_Info_2          object
Product_Info_3           int64
Product_Info_4         float64
Product_Info_5           int64
Product_Info_6           int64
Product_Info_7           int64
Ins_Age                float64
Ht                     float64
Wt                     float64
BMI                    float64
Employment_Info_1      float64
Employment_Info_2        int64
Employment_Info_3        int64
Employment_Info_4      float64
Employment_Info_5        int64
Employment_Info_6      float64
InsuredInfo_1            int64
InsuredInfo_2            int64
InsuredInfo_3            int64
InsuredInfo_4            int64
InsuredInfo_5            int64
InsuredInfo_6            int64
InsuredInfo_7            int64
Insurance_History_1      int64
Insurance_History_2      int64
Insurance_History_3      int64
Insurance_History_4      int64
Insurance_His

Id  Product_Info_1 Product_Info_2  Product_Info_3  Product_Info_4  \
0   1               1             D3              26        0.487179   
1   3               1             A2              26        0.076923   
2   4               1             D3              26        0.144667   
3   9               1             A1              26        0.151709   
4  12               1             A1              26        0.076923   

   Product_Info_5  Product_Info_6  Product_Info_7   Ins_Age        Ht  \
0               2               3               1  0.611940  0.781818   
1               2               3               1  0.626866  0.727273   
2               2               3               1  0.582090  0.709091   
3               2               1               1  0.522388  0.654545   
4               2               3               1  0.298507  0.672727   

         Wt       BMI  Employment_Info_1  Employment_Info_2  \
0  0.338912  0.472262              0.150                  3   
1  0.311715  0.484984              0.000                  1   
2  0.320084  0.519103              0.143                  9   
3  0.267782  0.486962              0.210                  9   
4  0.246862  0.428718              0.085                  9   

   Employment_Info_3  Employment_Info_4  Employment_Info_5  Employment_Info_6  \
0                  1               0.00                  2               0.50   
1                  3               0.07                  2               0.20   
2                  1               0.00                  2               0.45   
3                  1               0.00                  2               1.00   
4                  1               0.00                  2               0.20   

   InsuredInfo_1  InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0              2              2             11              3              1   
1              1              2              8              3              1   
2              1              2              3              3              1   
3              2              2              3              3              1   
4              1              2              8              3              1   

   InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
0              1              1                    2                    1   
1              1              1                    1                    1   
2              1              1                    2                    1   
3              1              1                    1                    1   
4              2              1                    2                    1   

   Insurance_History_3  Insurance_History_4  Insurance_History_5  \
0                    1                    3                  NaN   
1                    3                    1             0.001667   
2                    1                    3                  NaN   
3                    3                    1             0.000667   
4                    1                    3                  NaN   

   Insurance_History_7  Insurance_History_8  Insurance_History_9  \
0                    3                    2                    3   
1                    1                    1                    2   
2                    3                    2                    3   
3                    2                    1                    2   
4                    3                    2                    3   

   Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  Family_Hist_5  \
0              3            NaN       0.627451       0.760563            NaN   
1              2            NaN       0.529412       0.746479            NaN   
2              3       0.666667            NaN       0.661972            NaN   
3              2            NaN       0.686275       0.676056            NaN   
4              2       0.449275            NaN       0.380282            NaN   

   Medical_History_1  


Shape of test data : (19765, 127)

=============Data types in test data==============
Id                       int64
Product_Info_1           int64
Product_Info_2          object
Product_Info_3           int64
Product_Info_4         float64
Product_Info_5           int64
Product_Info_6           int64
Product_Info_7           int64
Ins_Age                float64
Ht                     float64
Wt                     float64
BMI                    float64
Employment_Info_1      float64
Employment_Info_2        int64
Employment_Info_3        int64
Employment_Info_4      float64
Employment_Info_5        int64
Employment_Info_6      float64
InsuredInfo_1            int64
InsuredInfo_2            int64
InsuredInfo_3            int64
InsuredInfo_4            int64
InsuredInfo_5            int64
InsuredInfo_6            int64
InsuredInfo_7            int64
Insurance_History_1      int64
Insurance_History_2      int64
Insurance_History_3      int64
Insurance_History_4      int64
Insurance_Hist

**From above and the data description provided by kaggle, we note the following :**

- Train dataset has 59,381 records and 128 features. Test dataset has 19,765 records with just 127 features (test data lack the `response` feature). Based on the data description, the `response` variable is an ordinal measure of risk at 8 levels. 


- Datasets comprise hundred variables describing the attributes of life insurance applicants, which have been anonymized. There are some missing values, but these will be looked at in details, later below at preliminary review with pandas profiling.


- `Product_Info_1-7` is the set of normalized variables related to products applied for. Only `Product_Info_2` is a string while the rest of the features are either integers or floats. Based on the data description, `Product_Info_2` is a nominal categorical variable so we will dummify this prior to modelling.


- `Ins_Age`, `Ht`, `Wt` and `BMI` have all been normalised and refer to the age, height, weight and BMI of the respective applicants. These are continuous variables so if there are any missing values, imputation will be made based on the distribution.


- `Employment_Info_1-6` relates to the employment history of the respective applicants, which have also been normalized. `Employment_Info_1,4 and 6` are continuous variables, while the rest (ie. `Employment_Info_2,3, and 5`) are nominal categorical variables. This will affect imputation of missing values (if any) later on. 


- `Insured_Info_1-7` are normalized and all nominal categorical variables. 


- `Insurance_History_1-9` except `Insurance_History_5` are nominal categorical variables. `Insurance_History_5` is a continuous variable. The information relates to the insurance history of the applicant.


- `Family_Hist_1-5` relates to family history of the applicant, which have also been normalized. `Family_Hist_1` is a nominal categorical vairable, whereas the rest are continuous variables. 


- `Medical_History_1-9` relates to the medical hisotry of applicant (normalized). `Medical_History_1, 10, 15, 24, 32` are discrete continuous variables, while the rest are nominal categorical variables. 


- `Medical_Keyword_1-48` is a set of dummy variables related to the presence/absence of a medical keyword associated with the application. 

In [76]:
# function that drops Id field and checks
def drop_id_check(df):
    print(f"Shape before dropping : {df.shape}") #check shape
    df.drop(columns = 'Id', axis = 1, inplace = True) #drop Id column
    print(f"Shape after dropping : {df.shape}")
    return df.head()

In [77]:
%%time
drop_id_check(train)

Shape before dropping : (59381, 128)
Shape after dropping : (59381, 127)
Wall time: 18 ms


Product_Info_1 Product_Info_2  Product_Info_3  Product_Info_4  \
0               1             D3              10        0.076923   
1               1             A1              26        0.076923   
2               1             E1              26        0.076923   
3               1             D4              10        0.487179   
4               1             D2              26        0.230769   

   Product_Info_5  Product_Info_6  Product_Info_7   Ins_Age        Ht  \
0               2               1               1  0.641791  0.581818   
1               2               3               1  0.059701  0.600000   
2               2               3               1  0.029851  0.745455   
3               2               3               1  0.164179  0.672727   
4               2               3               1  0.417910  0.654545   

         Wt       BMI  Employment_Info_1  Employment_Info_2  \
0  0.148536  0.323008              0.028                 12   
1  0.131799  0.272288              0.000                  1   
2  0.288703  0.428780              0.030                  9   
3  0.205021  0.352438              0.042                  9   
4  0.234310  0.424046              0.027                  9   

   Employment_Info_3  Employment_Info_4  Employment_Info_5  Employment_Info_6  \
0                  1                0.0                  3                NaN   
1                  3                0.0                  2             0.0018   
2                  1                0.0                  2             0.0300   
3                  1                0.0                  3             0.2000   
4                  1                0.0                  2             0.0500   

   InsuredInfo_1  InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0              1              2              6              3              1   
1              1              2              6              3              1   
2              1              2              8              3              1   
3              2              2              8              3              1   
4              1              2              6              3              1   

   InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
0              2              1                    1                    1   
1              2              1                    2                    1   
2              1              1                    2                    1   
3              2              1                    2                    1   
4              2              1                    2                    1   

   Insurance_History_3  Insurance_History_4  Insurance_History_5  \
0                    3                    1             0.000667   
1                    3                    1             0.000133   
2                    1                    3                  NaN   
3                    1                    3                  NaN   
4                    1                    3                  NaN   

   Insurance_History_7  Insurance_History_8  Insurance_History_9  \
0                    1                    1                    2   
1                    1                    3                    2   
2                    3                    2                    3   
3                    3                    2                    3   
4                    3                    2                    3   

   Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  Family_Hist_5  \
0              2            NaN       0.598039            NaN       0.526786   
1              2       0.188406            NaN       0.084507            NaN   
2              3       0.304348            NaN       0.225352            NaN   
3              3       0.420290            NaN       0.352113            NaN   
4              2       0.463768            NaN       0.408451            NaN   

   Medical_History_1  Medical_History_2  Medic

In [78]:
%%time
drop_id_check(test)

Shape before dropping : (19765, 127)
Shape after dropping : (19765, 126)
Wall time: 5.98 ms


Product_Info_1 Product_Info_2  Product_Info_3  Product_Info_4  \
0               1             D3              26        0.487179   
1               1             A2              26        0.076923   
2               1             D3              26        0.144667   
3               1             A1              26        0.151709   
4               1             A1              26        0.076923   

   Product_Info_5  Product_Info_6  Product_Info_7   Ins_Age        Ht  \
0               2               3               1  0.611940  0.781818   
1               2               3               1  0.626866  0.727273   
2               2               3               1  0.582090  0.709091   
3               2               1               1  0.522388  0.654545   
4               2               3               1  0.298507  0.672727   

         Wt       BMI  Employment_Info_1  Employment_Info_2  \
0  0.338912  0.472262              0.150                  3   
1  0.311715  0.484984              0.000                  1   
2  0.320084  0.519103              0.143                  9   
3  0.267782  0.486962              0.210                  9   
4  0.246862  0.428718              0.085                  9   

   Employment_Info_3  Employment_Info_4  Employment_Info_5  Employment_Info_6  \
0                  1               0.00                  2               0.50   
1                  3               0.07                  2               0.20   
2                  1               0.00                  2               0.45   
3                  1               0.00                  2               1.00   
4                  1               0.00                  2               0.20   

   InsuredInfo_1  InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0              2              2             11              3              1   
1              1              2              8              3              1   
2              1              2              3              3              1   
3              2              2              3              3              1   
4              1              2              8              3              1   

   InsuredInfo_6  InsuredInfo_7  Insurance_History_1  Insurance_History_2  \
0              1              1                    2                    1   
1              1              1                    1                    1   
2              1              1                    2                    1   
3              1              1                    1                    1   
4              2              1                    2                    1   

   Insurance_History_3  Insurance_History_4  Insurance_History_5  \
0                    1                    3                  NaN   
1                    3                    1             0.001667   
2                    1                    3                  NaN   
3                    3                    1             0.000667   
4                    1                    3                  NaN   

   Insurance_History_7  Insurance_History_8  Insurance_History_9  \
0                    3                    2                    3   
1                    1                    1                    2   
2                    3                    2                    3   
3                    2                    1                    2   
4                    3                    2                    3   

   Family_Hist_1  Family_Hist_2  Family_Hist_3  Family_Hist_4  Family_Hist_5  \
0              3            NaN       0.627451       0.760563            NaN   
1              2            NaN       0.529412       0.746479            NaN   
2              3       0.666667            NaN       0.661972            NaN   
3              2            NaN       0.686275       0.676056            NaN   
4              2       0.449275            NaN       0.380282            NaN   

   Medical_History_1  Medical_History_2  Medic

In [79]:
# function that creates dummies
def pdt_info_2_dummies(df):
    print(f"Shape before dummifying : {df.shape}") #check shape
    df = pd.get_dummies(df, prefix = 'Product_Info_2', columns = ['Product_Info_2'])
    print(f"Shape after dummifying : {df.shape}")
    return df

In [80]:
%%time
train = pdt_info_2_dummies(train)

Shape before dummifying : (59381, 127)
Shape after dummifying : (59381, 145)
Wall time: 43.2 ms


In [81]:
%%time
test = pdt_info_2_dummies(test)

Shape before dummifying : (19765, 126)
Shape after dummifying : (19765, 144)
Wall time: 17 ms


### Detailed review
Using pandas profiling report, a review of the data is performed. 

In [55]:
# generate profile report
train_report = ProfileReport(train)
train_report.to_widgets()

Based on the pandas profiling report for training set:

- `Product_Info_1` only has values 1 or 2 with a large majority(about 99%)of its records showing 1. There are no missing values. `Product_Info_3` has 34 distinct values. This field is a nominal categorical variable with number 26 being the majority (85%), with other commin numbers being 10, 29, 31 and 15. There are no missing values. `Product_Info_5` has 2 distinct values, 2 and 3, with 2 being the large majority ( about 99%). There are no missing values. `Product_Info_6` has 2 distinct values, 3 and 1, with 3 being the large majority (about 79%). There are no missing values. `Product_Info_7` has 3 distinct values, 1, 3 and 2, with 1 being the large majority (about 98%), followed by 3 (2.18%). There are no missing values. 


- `Product_Info_4` has 1491 distinct values, including zero (0.017%). There are no missing values. This field is a continuous variable, with most common values at 0.00769, 0.231 (rounded to 3sf).


- `Ins_Age` is a continuous variable that ranges from 0-1 (as it has been normalized). It has 65 distinct values, and no missing values. Average value is at 0.406 (3sf). It appears to have a bimodal distribution, which suggests that the records comprise applicants ages somewhere below the middle age (peak at about 0.2) and somewhere above middle age (peak at about 0.5). `Ht` is also normalized (values ranging from 0-1). There are no missing values. It distributed with a mean of 0.707 and median of 0.709. `Wt` is also normalized (values ranging from 0-1). There are no missing values. It appears to have a right skewed distribution, with a mean of 0.292 and median of 0.289. `BMI` has a right skewed distribution with mean of 0.469, median of 0.451. There are no missing values. 


- `Employment_Info_1` has  right skewed distribution with mean at 0.0776 and median at 0.06. It has 19 missing values. **As this is a continuous variable with skewed distribution and small % of missing values, the missing values will be imputed with the median value.** `Employment_Info_4` has 11.4% missing values, with a right skewed distribution, where median and mode is 0 (about 74%). **Missing values in this column will be imputed with the median (zero).** `Employment_Info_6`has 18.3% missing values, with a trimodel distribution. It has a mean of 0.36, median of 0.25 and mode of 1. **As there is a trimodel distribution, and therefore hard to decide between which value (ie. the mean, median or mode) is most appropriate, I will use sklearn's IterativeImputer to impute the missing values in this field.**


- `Employment_Info_2` is a discrete continuous variable, with a mode of 9. It has a bimodal distribution, and no missing values. `Employment_Info_3` only has 2 distinct values (1 or 3). Majority of the records have number 1 (about 85%). There are no missing values. `Employment_Info_5` has distinct values 2 or 3, and no missing values. 85% of the records show value 2. 


- `InsuredInfo_1` has values 1,2 and 3, with mode of 1 (about 78%) and no missing values. `InsuredInfo_2` has values 2 and 3, and mode of 2 (about 98%). `InsuredInfo_3` has 11 distinct values, and no missing values. It has a bimodal distribution, with most common values being firstly 8 and then 3. `InsuredInfo_4` has 2 distinct values with 3 being the mode (about 88%) and no missing values. `InsuredInfo_5` has values 1 and 3, with 1 being the majority (about 97%) and no missing values. `InsuredInfo_6` has values 1 and 2, in relatively equal proportions and no missing values. `InsuredInfo_7`has majority values 1 (about 97%) and 3, and no missing values. 


- `Insurance_History_1` has distinct values 2 and 1 (with 2 being the majority of about 72%) and no missing values. `Insurance_History_2` has distinct values 1,2 and 3. 1 is the mode (96%) and there are no missing values. `Insurance_History_3` has 1,2 and 3 distinct values too, with 3 being the mode. There are no missing values here as well. `Insurance_History_4`has a mode of 1 and no missing values. `Insurance_History_7`has a mode of 1 and no missing values.`Insurance_History_8`and `Insurance_History_9` has a mode of 2 and no missing values.


- `Insurance_History_5` is a continuous variable with values all very close to zero. Its mode is 0. **It has 42% missing values, which could be imputed with the mode, zero.**


- `Family_Hist_1` has distinct values 1, 2 and 3 with 3 being the mode, and no missing values. 


- `Family_Hist_2` has a relatively normal distribution (slightly right skewed), with 48% missing values. Based on the distribution, these can be **imputed with median.** `Family_Hist_3` has a relatively left-skewed distribution, with 57.7% missing values. **Missing values can be imputed with median as well.** `Family_Hist_4` has slightly right skewed distribution with 32% missing values, which can be **imputed with the median.**


- `Medical_History_1, 10, 15, 24 and 32` are discrete continuous variables. `Medical_History_1` has a mode of 1 and a heavily right skewed distribution. It has 15% missing values which we will **impute with the mode, 1.**


- The rest of features of medical history has nominal categorical variables with no missing values. Most are characterised by right skewed distribution. 


#### Significant missing values (>70%) : 

- `Family_Hist_5` has >70% missing values, and will be dropped. 
- `Medical_History_10` has 99% missing variables so will be dropped. 
- `Medical_History_15` has 75.1% missing variables so will be dropped. 
- `Medical_History_24` has 93% missing variables so will be dropped. 
- `Medical_History_32` has 98% missing variables so will be dropped. 

In [56]:
# generate profile report
test_report = ProfileReport(test)
test_report.to_widgets()

Based on pandas profiling report for test set :
- There is the exact same features on the train and test set except that test set does not have a response column (which is the target variable). So anything that will be dropped in the train set, will have to be dropped in the test set too. 


- **Missing values for in the test set will be imputed based on values in the train set.** No information of the test set should be inferred from as that would be data leakage. 

  - `Employment_Info_1` to be imputed with median value of the train set
  - `Employment_Info_4` to be imputed with median value of the train set
  - `Employment_Info_6` to be imputed with the median value of the train set (we can't apply the IterativeImputer here) 
  - `Insurance_History_5`to be imputed with zero
  - `Family_Hist_2`to be imputed with median value of the train set
  - `Family_Hist_3`to be imputed with median value of the train set, same with `Family_Hist_4`
  - `Medical_History_1` to be imputed with 1.
  
  
- The following columns will be dropped based on train set data :

  - `Family_Hist_5` 
  - `Medical_History_10` 
  - `Medical_History_15` 
  - `Medical_History_24` 
  - `Medical_History_32` 

### Dealing with missing values
I'll drop the features abovementioned, and perform the relevant imputations here. 

In [82]:
# function that gets the percentage of missing values 
def deal_missing(df):
    '''drop columns with significant missing values and impute other features with missing values'''
    
    #set columns to drop
    cols_to_drop = ['Family_Hist_5','Medical_History_10','Medical_History_15','Medical_History_24','Medical_History_32'] 
    
    #drop columns
    df.drop(columns = cols_to_drop, axis = 1, inplace = True) 
    
    # impute missing values for the rest of the features with missing values 
    df['Employment_Info_1'].fillna((train['Employment_Info_1'].median()), inplace = True) 
    df['Employment_Info_4'].fillna((train['Employment_Info_4'].median()), inplace = True) 
    df['Employment_Info_6'].fillna((train['Employment_Info_6'].median()), inplace = True) 
    df['Insurance_History_5'].fillna(0, inplace = True)
    df['Family_Hist_2'].fillna((train['Family_Hist_2'].median()), inplace = True)
    df['Family_Hist_3'].fillna((train['Family_Hist_3'].median()), inplace = True)
    df['Family_Hist_4'].fillna((train['Family_Hist_4'].median()), inplace = True)
    df['Medical_History_1'].fillna(1, inplace = True)    
    
    #check
    print(f"Check if any more missing values : \n{df.isnull().sum()}")
    print()
    
    return df

In [83]:
%%time
#apply function on train set 
deal_missing(train)

Check if any more missing values : 
Product_Info_1         0
Product_Info_3         0
Product_Info_4         0
Product_Info_5         0
Product_Info_6         0
Product_Info_7         0
Ins_Age                0
Ht                     0
Wt                     0
BMI                    0
Employment_Info_1      0
Employment_Info_2      0
Employment_Info_3      0
Employment_Info_4      0
Employment_Info_5      0
Employment_Info_6      0
InsuredInfo_1          0
InsuredInfo_2          0
InsuredInfo_3          0
InsuredInfo_4          0
InsuredInfo_5          0
InsuredInfo_6          0
InsuredInfo_7          0
Insurance_History_1    0
Insurance_History_2    0
Insurance_History_3    0
Insurance_History_4    0
Insurance_History_5    0
Insurance_History_7    0
Insurance_History_8    0
Insurance_History_9    0
Family_Hist_1          0
Family_Hist_2          0
Family_Hist_3          0
Family_Hist_4          0
Medical_History_1      0
Medical_History_2      0
Medical_History_3      0
Medical_Histor

Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
0                   1              10        0.076923               2   
1                   1              26        0.076923               2   
2                   1              26        0.076923               2   
3                   1              10        0.487179               2   
4                   1              26        0.230769               2   
...               ...             ...             ...             ...   
59376               1              10        0.230769               2   
59377               1              26        0.230769               2   
59378               1              26        0.076923               2   
59379               1              10        0.230769               2   
59380               1              26        0.076923               2   

       Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
0                   1               1  0.641791  0.581818  0.148536  0.323008   
1                   3               1  0.059701  0.600000  0.131799  0.272288   
2                   3               1  0.029851  0.745455  0.288703  0.428780   
3                   3               1  0.164179  0.672727  0.205021  0.352438   
4                   3               1  0.417910  0.654545  0.234310  0.424046   
...               ...             ...       ...       ...       ...       ...   
59376               3               1  0.074627  0.709091  0.320084  0.519103   
59377               3               1  0.432836  0.800000  0.403766  0.551119   
59378               3               1  0.104478  0.745455  0.246862  0.360969   
59379               3               1  0.507463  0.690909  0.276151  0.462452   
59380               3               1  0.447761  0.781818  0.382845  0.539563   

       Employment_Info_1  Employment_Info_2  Employment_Info_3  \
0                  0.028                 12                  1   
1                  0.000                  1                  3   
2                  0.030                  9                  1   
3                  0.042                  9                  1   
4                  0.027                  9                  1   
...                  ...                ...                ...   
59376              0.020                  1                  3   
59377              0.100                  9                  1   
59378              0.035                  9                  1   
59379              0.038                  9                  1   
59380              0.123                  9                  1   

       Employment_Info_4  Employment_Info_5  Employment_Info_6  InsuredInfo_1  \
0                0.00000                  3             0.2500              1   
1                0.00000                  2             0.0018              1   
2                0.00000                  2             0.0300              1   
3                0.00000                  3             0.2000              2   
4                0.00000                  2             0.0500              1   
...                  ...                ...                ...            ...   
59376            0.00000                  3             0.0250              1   
59377            0.00001                  2             0.3500              1   
59378            0.00000                  2             0.2500              1   
59379            0.00000                  3             0.2500              1   
59380            0.00000                  2             0.3000              1   

       InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0                  2              6              3              1   
1                  2              6              3              1   
2                  2              8              3              1   
3                  2              8              3              1   
4                  2              6           

In [84]:
%%time
deal_missing(test)

Check if any more missing values : 
Product_Info_1         0
Product_Info_3         0
Product_Info_4         0
Product_Info_5         0
Product_Info_6         0
Product_Info_7         0
Ins_Age                0
Ht                     0
Wt                     0
BMI                    0
Employment_Info_1      0
Employment_Info_2      0
Employment_Info_3      0
Employment_Info_4      0
Employment_Info_5      0
Employment_Info_6      0
InsuredInfo_1          0
InsuredInfo_2          0
InsuredInfo_3          0
InsuredInfo_4          0
InsuredInfo_5          0
InsuredInfo_6          0
InsuredInfo_7          0
Insurance_History_1    0
Insurance_History_2    0
Insurance_History_3    0
Insurance_History_4    0
Insurance_History_5    0
Insurance_History_7    0
Insurance_History_8    0
Insurance_History_9    0
Family_Hist_1          0
Family_Hist_2          0
Family_Hist_3          0
Family_Hist_4          0
Medical_History_1      0
Medical_History_2      0
Medical_History_3      0
Medical_Histor

Product_Info_1  Product_Info_3  Product_Info_4  Product_Info_5  \
0                   1              26        0.487179               2   
1                   1              26        0.076923               2   
2                   1              26        0.144667               2   
3                   1              26        0.151709               2   
4                   1              26        0.076923               2   
...               ...             ...             ...             ...   
19760               1              26        0.128205               2   
19761               1              10        0.230769               2   
19762               1              26        0.076923               2   
19763               1              26        0.487179               2   
19764               1              26        1.000000               2   

       Product_Info_6  Product_Info_7   Ins_Age        Ht        Wt       BMI  \
0                   3               1  0.611940  0.781818  0.338912  0.472262   
1                   3               1  0.626866  0.727273  0.311715  0.484984   
2                   3               1  0.582090  0.709091  0.320084  0.519103   
3                   1               1  0.522388  0.654545  0.267782  0.486962   
4                   3               1  0.298507  0.672727  0.246862  0.428718   
...               ...             ...       ...       ...       ...       ...   
19760               3               1  0.537313  0.636364  0.152720  0.284929   
19761               3               1  0.134328  0.618182  0.225941  0.446361   
19762               3               1  0.850746  0.727273  0.414226  0.655973   
19763               3               1  0.522388  0.745455  0.529289  0.818698   
19764               1               1  0.343284  0.781818  0.382845  0.539563   

       Employment_Info_1  Employment_Info_2  Employment_Info_3  \
0                  0.150                  3                  1   
1                  0.000                  1                  3   
2                  0.143                  9                  1   
3                  0.210                  9                  1   
4                  0.085                  9                  1   
...                  ...                ...                ...   
19760              0.065                  9                  1   
19761              0.015                 14                  1   
19762              0.051                  1                  3   
19763              0.065                 14                  1   
19764              0.325                 15                  1   

       Employment_Info_4  Employment_Info_5  Employment_Info_6  InsuredInfo_1  \
0                   0.00                  2               0.50              2   
1                   0.07                  2               0.20              1   
2                   0.00                  2               0.45              1   
3                   0.00                  2               1.00              2   
4                   0.00                  2               0.20              1   
...                  ...                ...                ...            ...   
19760               0.00                  2               0.30              2   
19761               0.00                  3               0.00              1   
19762               0.00                  2               0.50              1   
19763               0.00                  2               0.10              1   
19764               0.05                  2               1.00              1   

       InsuredInfo_2  InsuredInfo_3  InsuredInfo_4  InsuredInfo_5  \
0                  2             11              3              1   
1                  2              8              3              1   
2                  2              3              3              1   
3                  2              3              3              1   
4                  2              8           

### Save cleaned data

In [85]:
train_clean = train.to_csv("../data/train-clean.csv", index = False)
test_clean = test.to_csv("../data/test-clean.csv", index = False)

There are no more missing values and data is cleaned! Refer to [this notebook](02-eda.ipynb) for data exploration.